### DISCUSSION TITLE
Why SWR always assert the data may be undefined?

### DISCUSSION THREAD

**Discussion by cw1997** [vercel/swr#2942] [Q&A]
The sample code at here:
````
const {data, error, isLoading} = useSWR(['passport/session/get'], () => get_session())
if (!isLoading && error == null) {
  console.log(data.user_id)
}
````

![image](https://github.com/vercel/swr/assets/11588555/51a5a641-216d-4d32-9deb-2eb743e7fbe5)

In my opinion, I think if `isLoading` is false that means fetcher is executed finished, and if `error` is null, that means the executed result is successful, so data is must not undefined.

But I see the TypeScript report the data may be undefined in this case, is there any way to optimize the logic of typing narrow? Actually, I think in this condition, SWR should assert the data must not be undefined, is this an issue that can be improved?

Thank you very much!

**Comment by GoldenChrysus** [comment]
Isn't this just a TypeScript LSP concern rather than an SWR one? You haven't actually confirmed that `data` isn't null or undefined in that block, nor guaranteed to SWR that `data` will always have a value, so the LSP is warning you. The LSP isn't aware of the runtime relationship between `isLoading`, `data`, and `error`. Without a guarantee, `data` itself definitely must allow for empty values considering there will be no data before the request to get the data.

`data` being guaranteed to have data means you already have data to provide (`fallbackData` or use `suspense: true`) per the SWR typing:
```
interface SWRResponse<Data = any, Error = any, Config = any> {
    /**
     * The returned data of the fetcher function.
     */
    data: BlockingData<Data, Config> extends true ? Data : Data | undefined;
    ...
}
```
```
type BlockingData<Data = any, Options = SWROptions<Data>> = Options extends undefined ? false : Options extends {
    suspense: true;
} ? true : Options extends {
    fallbackData: Data;
} ? true : false;
```

Therefore a simple example like this guarantees that `data` has a value and is no longer `undefined` or `null` in your LSP.

```
interface MyConfig extends SWRConfiguration {
    fallbackData: string,
}

const { data } = useSWR<string, string, MyConfig>('hello', () => 'b', {
    fallbackData: 'a',
});
```

However, you have not provided SWR with any initial data or a suspense guarantee, so from SWR's perspective there is every chance you will not receive any data from your fetcher and thus have no data to serve.

Otherwise, a simple check that `data` is actually present will do.
